aten/src/ATen/native/cuda/GridSampler.cu

```diff
@@ -0,0 +1,782 @@
+#include "ATen/ATen.h"
+#include "ATen/native/GridSampler.h"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/detail/TensorInfo.cuh"
+#include "ATen/cuda/detail/IndexUtils.cuh"
+#include "ATen/cuda/detail/KernelUtils.h"
+
+namespace at { namespace native {
+
+using namespace at::cuda::detail;
+
+using at::native::detail::GridSamplerInterpolation;
+using at::native::detail::GridSamplerPadding;
+
+namespace {
+	static __forceinline__ __device__
+	int clip_coordinates(int in, int clip_limit) {
+	  return ::min(clip_limit - 1, ::max(in, static_cast<int>(0)));
+	}
+
+	static __forceinline__ __device__
+	bool within_bounds_2d(int h, int w, int H, int W) {
+	  return h >= 0 && h < H && w >= 0 && w < W;
+	}
+
+	static __forceinline__ __device__
+	bool within_bounds_3d(int d, int h, int w, int D, int H, int W) {
+	  return d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W;
+	}
+
+  template<typename scalar_t>
+	static __forceinline__ __device__
+  void safe_add_2d(scalar_t *data, int h, int w,
+                   int sH, int sW, int H, int W,
+                   scalar_t delta) {
+    if (h >= 0 && h < H && w >= 0 && w < W) {
+      atomicAdd(data + h * sH + w * sW, delta);
+    }
+  }
+
+  template<typename scalar_t>
+	static __forceinline__ __device__
+  void safe_add_3d(scalar_t *data, int d, int h, int w,
+                   int sD, int sH, int sW, int D, int H, int W,
+                   scalar_t delta) {
+    if (d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W) {
+      atomicAdd(data + d * sD + h * sH + w * sW, delta);
+    }
+  }
+
+	template <typename scalar_t>
+	__launch_bounds__(1024)
+	__global__ void grid_sampler_2d_kernel(
+	    const int nthreads,
+	    TensorInfo<scalar_t, int> input,
+	    TensorInfo<scalar_t, int> grid,
+	    TensorInfo<scalar_t, int> output,
+	    const GridSamplerPadding padding_mode) {
+
+	  int C = input.sizes[1];
+	  int inp_H = input.sizes[2];
+	  int inp_W = input.sizes[3];
+	  int out_H = grid.sizes[1];
+	  int out_W = grid.sizes[2];
+	  int inp_sN = input.strides[0];
+	  int inp_sC = input.strides[1];
+	  int inp_sH = input.strides[2];
+	  int inp_sW = input.strides[3];
+	  int grid_sN = grid.strides[0];
+	  int grid_sH = grid.strides[1];
+	  int grid_sW = grid.strides[2];
+	  int grid_sCoor = grid.strides[3];
+	  int out_sN = output.strides[0];
+	  int out_sC = output.strides[1];
+	  int out_sH = output.strides[2];
+	  int out_sW = output.strides[3];
+
+	  CUDA_KERNEL_LOOP(index, nthreads) {
+	  	const int w = index % out_W;
+	  	const int h = (index / out_W) % out_H;
+	  	const int n = index / (out_H * out_W);
```

Comment:
In THNN, this used to compute `w, h, n` as if `index` is the flattened index of a tensor with shape `[W x H x N]`. I changed so it is the flattened index of tensor w/ shape `[N x H x W]`. This allows me to take advantage of that `grad_out` is contiguous in backward, e.g., on line 439.

Reply:
I don't understand this comment. Did you change the output format of the tensor? If not, how can the choice of w/h/n here affect line 439? You're just running the forward kernel and writing the results to output; if you didn't change the output format, this choice shouldn't affect how you do computation in backward.

Reply:
Maybe what you meant is, you uniformly changed the index convention for all kernels; it doesn't matter for forwards but it does matter for backwards (where you can avoid an accumulation)?

Reply:
Sorry for not being clear. Here is what I meant. 

In THCUNN, the conversion from `index` to `w, h, n` is s.t. `index = w * (out_H * out_N) + h * out_N + n`, i.e., incrementing `index` by `1` correspond to incrementing `n` by `1`. But `n` is the outermost dimension, so I changed the conversion s.t. `index = n * (out_W * out_H) + h * out_W + w`. This way, to get the current `w`-slice of a contiguous tensor, one can just do `tensor_ptr_NHW = tensor.data + index * tensor_sW`. 

Yes, this doesn't matter for forward. But I made the change for both forward and backward for consistency.